### PR TITLE
chore(permissions and service): Add permission getters and requests f…

### LINKED ISSUE #12: Create logic for location permission and service
Add use cases for querying the location permission status and asking for permission grant
Add use cases for querying and enabling location service

### PR DESCRIPTION
**Pull request by ronw98**
…or notification and location

Add location service getter and request

closes #12

### CODE REVIEW

File: android/app/build.gradle
Comment (BriceCroix) [inline android/app/build.gradle:1]: What is the purpose of these lines ?

File: lib/domain/entities/location_service_status.dart
Comment (BriceCroix) [inline lib/domain/entities/location_service_status.dart:1]: Please add documentations, I am unsure what the purpose of this class is.

File: lib/domain/entities/notification_config.dart
Comment (BriceCroix) [inline lib/domain/entities/notification_config.dart:1]: Please add documentations, I am unsure what the purpose of this class is.

File: lib/domain/usecases/permission_usecases/handle_location_permission.dart [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+///
+/// If this status is other than [SystemPermissionStatus.permanentlyDenied],
+/// shows a dialog for the user to grant the permission and returns the
+/// resulting permission status.
+///
+/// If the initial status is [SystemPermissionStatus.permanentlyDenied] or a
+/// granted status to some degree, this usecase only returns the status and does
+/// nothing else.
+@injectable
+class HandleLocationPermission
```
Comment (BriceCroix) [inline lib/domain/usecases/permission_usecases/handle_location_permission.dart:19]: `Handle` is way too generic, and it is usually a good choice to name variables and classes differently. Maybe `Request` or `Ensure` there ?

File: lib/main.dart
Comment (BriceCroix) [inline lib/main.dart:1]: I daisagree with that, permissions should be asked whenever necessary, is it possible to move these request in the (future) activity screens ?

File: lib/data/datasources/position_source.dart [hunk 2]
```diff
@@ -8,23 +9,49 @@ abstract class PositionSource {
...
         // Set to best available accuracy.
         accuracy: LocationAccuracy.best,
+        intervalDuration: const Duration(milliseconds: 500),
+        foregroundNotificationConfig: notificationConfig,
       ),
     );
   }
+
+  @override
+  Future<bool> isLocationServiceEnabled() {
```
Comment (BriceCroix) [inline lib/data/datasources/position_source.dart:49]: What is the difference between `LocationService` there and `locationPermission` in the Permission source ?

File: lib/data/datasources/position_source.dart [hunk 3]
```diff
@@ -8,23 +9,49 @@ abstract class PositionSource {
...
   }
 
   @override
-  Stream<Position> getPositionStream() {
+  Stream<Position> getPositionStream(
+    ForegroundNotificationConfig notificationConfig,
+  ) {
     return Geolocator.getPositionStream(
-      locationSettings: const LocationSettings(
+      locationSettings: AndroidSettings(
```
[the remark this answers is not in the document]
Comment (ronw98) [reply lib/data/datasources/position_source.dart:39]: It is supposed to matter and it does.
Below two images with the interval between two location stream updates
| With `AndroidSettings` and interval duration set to 500ms | With `LocationSettings` |
| ----- | ------ |
image | image |

Sometimes platform specific code is unavoidable, especially when calling native services that may take different arguments. Although I should have written something like below to only use AndroidSettings when necessary.
I can remove the `intervalDuration` setting if you want but we'll need to add it at some point and we need the AndroidSettings' `foregroundNotificationConfig` anyway to get updates in the background
```dart
      locationSettings: Platform.isAndroid
          ? AndroidSettings(
              // Set to best available accuracy.
              accuracy: LocationAccuracy.best,
              intervalDuration: const Duration(milliseconds: 500),
              foregroundNotificationConfig: notificationConfig,
            )
          : const LocationSettings(
              accuracy: LocationAccuracy.best,
            ),
```

[the remark this answers is not in the document]
Comment (BriceCroix) [reply lib/data/datasources/position_source.dart:39]: Yes that proposal seems perfect ! Perhaps 1 second would be enough though, you could add a global constant at the top of the file for that purpose.

File: android/app/build.gradle
Comment (ronw98) [reply android/app/build.gradle:1]: Add gms dependencies for location, we could create a version of the app that does not use GMS in the future if we want to publish on f-droid

File: lib/domain/entities/location_service_status.dart
Comment (ronw98) [reply lib/domain/entities/location_service_status.dart:1]: done

File: lib/domain/entities/notification_config.dart
Comment (ronw98) [reply lib/domain/entities/notification_config.dart:1]: done

File: lib/data/datasources/position_source.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronw98) [reply lib/data/datasources/position_source.dart:39]: done

File: lib/data/datasources/position_source.dart (hunk 2 above)
Comment (ronw98) [reply lib/data/datasources/position_source.dart:49]: location permission is whether or not the app is authorized to access the device location
`LocationService` is the system service that manages the user's location, this service can be enabled or disabled by the user in the top system bar

File: lib/domain/usecases/permission_usecases/handle_location_permission.dart (hunk 1 above)
Comment (ronw98) [reply lib/domain/usecases/permission_usecases/handle_location_permission.dart:19]: i know handle is generic but this use-case does not necessarily requests the permission, it does so only if the permission is not granted and not permanently denied

So neither `ensure` not `request` works well

File: lib/main.dart
Comment (ronw98) [reply lib/main.dart:1]: yup sure, this was mainly for testing / showcase purposes, this can definitely move somehere else

**Pull request merged**